### ISSUE TITLE
Check exhaustiveness of ToMtree

### ISSUE THREAD

**Issue by iTakeshi** [scalameta/paradise#43]
`ToMtree` lacks some converters necessary for dealing with some particular cases (as you can see in issues #35, #36, #37, #40, #41, #42). Even though this does NOT reduce the value of @Dveim 's awesome work at all, `ToMtree` should be exhaustive in converting `gtree`, because it is one of the core modules of this plugin.

So, how can/should we write specs which properly ensures the exhaustiveness?

One simple idea is hand-write every possible pair of `gtree` and converted `mtree`, but this will be another tremendous work and origin of human errors. Moreover, I (probably we) don't want to fix them by hand when the specification of `scala.reflect` or `scala.meta` is changed.

Any good ideas?

**Comment by DavidDudson** [comment]
On Idea might be to fuzzy test (Property test with large amounts of time and randomness) ToMTree. We can generate random bits of valid Scala syntax via the meta.Tree's and then call .show[Syntax] to get some sort of textual representation. We can then pipe that through. This is not exhaustive, but given enough iterations it would definitely find the missing parts. 

Note: There are a lot of known problems with ToMTree. The above would more be for when ToMTree  is presumed complete, just to find the edge cases where it does not work. ToMTree is going to be an ongoing problem for while. We just need to wait for people to get blocked, fix those issues, then carry on. The sheer amount of working in completing ToMTree is non trivial.

**Comment by Dveim** [comment]
Indeed, converter is not finished yet (I was adding missing parts by demand). In my option, best option would be to have test cases for all API from `scala.meta.Trees` and verify exhaustiveness manually, not sure if autochecking is worth the efforts (and, most importantly, is really hard to have guarantees for exhaustiveness, we can trap in "chicken-or-the-egg" problem).

**Comment by olafurpg** [comment]
I agree with @Dveim that a manually written test suite that covers all corners of scala.meta.Tree is a must-have.

I would also argue that it necessary to have a test suite that runs the converter on several million lines of code from open source projects and checks for properties like "the converter doesn't crash". My experience with scalafmt is that such tests are really effective at finding bugs that users otherwise bump into. This also helps prioritize for common coding pattern over rare corner cases.

I would be happy to work on infrastructure to run the converter on lots of open source code. I've already mentioned this to @xeno-by. I'm hoping to get some semantic info out of scala.meta to implement basic rewrites in [scalafix](https://github.com/scalacenter/scalafix), a migration tool to dotty.

**Comment by iTakeshi** [comment]
Thanks for your comments.
I understand manually-written specs are necessary, and additional tests on randomly-generated code or OSS projects will be helpful.

@Dveim are you already working on this? If there are something I can help with, I'd be happy to work on them.

**Comment by olafurpg** [comment]
FWIW, [here's](https://github.com/olafurpg/scalafmt/blob/master/core/src/test/scala/org/scalafmt/util/ScalaFile.scala) a zip file of 3 million lines of only scala sources. I use [this](https://github.com/olafurpg/scalafmt/blob/master/core/src/test/scala/org/scalafmt/util/ScalaFile.scala) code in scalafmt to automatically download the zip, uncompress it, parse the files and run analysis on them.

**Comment by Dveim** [comment]
@iTakeshi No, I'll start in October. Task is easily parallelizable, would be great to work on it together.

**Comment by iTakeshi** [comment]
@olafurpg That's a cool stuff!! So you're going to apply it to `ToMtree`, right? I'm really looking forward it!

@Dveim OK, then I'll try first on this weekend if I can.

**Comment by olafurpg** [comment]
@iTakeshi That is the plan : )

**Comment by iTakeshi** [comment]
I wrote a prototype of the test suite. https://github.com/iTakeshi/paradise/commit/6e1485b75e1d64965230ed5cdc0d8f102b84499a

There are some points:
* Is the style in which I wrote OK?
* Does anybody know how to import just `scala.reflect.api.Quasiquote` (i.e. `q""` syntax) without doing `import scala.reflect.runtime.universe._` ?
* My implementation still fails because `shouldEqual` compares `actual` and `expected` too strictly. It can be solved with an [`Equality`](http://doc.scalactic.org/3.0.0/index.html#org.scalactic.Equality) object to correctly compare them, so I'll do this from now.

Any comments or suggestions are welcome.

**Comment by olafurpg** [comment]
@iTakeshi Here's a different approach that might be worth considering, it looks roughly like this:

```scala
def check(code: String): Unit = {
  test(code) {
    val mTree: m.Tree = code.parse[Stat].get
    val gTree: g.Tree = <parse code into g.Tree>
    val convertedTree: m.Tree = convert(gTree)
    assert(mTree.structure == convertedTree.structure)
  }
}

check("println(42)")
check("function(a = 2)")
```

I haven't tried to implement it, but it would allow for very a concise syntax. Structural equality should be enough to ensure two trees are the same, [here](https://github.com/olafurpg/scalafmt/blob/db873b0eefa031a026edc48e76244e78b7f533e5/core/src/test/scala/org/scalafmt/util/FormatAssertions.scala#L54-L57) is a method I use to produce readable diffs when comparing two `.structure` outputs (note requires `"com.googlecode.java-diff-utils" % "diffutils" % "1.3.0"`).

**Comment by iTakeshi** [comment]
@olafurpg Nice solution. Thanks!
I added a commit with an implementation of your suggestion. How is that?

**Comment by olafurpg** [comment]
@iTakeshi That looks great 👍

**Comment by iTakeshi** [comment]
@olafurpg Thank you!

@Dveim How do you think? If you feel OK, I'll proceed to add other cases

**Comment by iTakeshi** [comment]
BTW, I think that checking equality (not structural, but semantic) of two trees is useful for solving #26, how do you think? cc @DavidDudson

**Comment by olafurpg** [comment]
@iTakeshi I think semantic equality is a research topic on it's own. Two different ASTs can be observationally equivalent, for example `def x = {2}` and `def x = 2`. Solving this generally is quite hard.

However, you can be sure that two trees with identical ASTs are semantically equal. It would be nice to have a `normalize: Tree => Tree` method in scala.meta that can do basic semantically preserving transformations. That would enable us to check if two different trees have the same normalized structure.

**Comment by Dveim** [comment]
@iTakeshi Very nice approach, I like it. One note -- there are few syntaxes (`inline`, `type A = B | C | D`, mb more) that wont be parsed correctly by `GParser` (or it should be adjusted). These should be handled separately, I think.

**Comment by DavidDudson** [comment]
@itakeshi it might actually work. Sounds good :+1:

**Comment by iTakeshi** [comment]
Thanks @olafurpg & @DavidDudson , I'll try that.

@Dveim Ah, I didn't notice those cases, they should be considered in near future. Anyway I submitted a PR with current implementation in order to just step forward. Please review that, and I hope working together.

**Comment by iTakeshi** [comment]
Solved with #69 , closing.
